<h1>Final Capstone Project</h1>

### This notebook is on Segmenting and Clustering Neighborhoods in Toronto - Week 3 of Capstone



# Table of Contents 

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1.  <a href="#item1">Week 3 - Section 1</a>

2.  <a href="#item2">Week 3: Section 2</a>

3.  <a href="#item3">Week 3: Section 3</a>

    </font>
    </div>

In [89]:
import pandas as pd
import numpy as np
print ("Assignments for Week 3 -- all in 1 Notebook")

Assignments for Week 3 -- all in 1 Notebook


## Week 3 - Section 1:
### Scraping from website, examining data, and pulling into dataframe:

In [90]:
import requests


from bs4 import BeautifulSoup



In [91]:
website = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
r = requests.get(website)
r

<Response [200]>

In [92]:
# examining the data, including structure of it.
html=r.content
soup = BeautifulSoup(html, 'html.parser')
#soup

In [93]:
#Grabs the first table in the website and converts to dataframe.
CanPostal = pd.read_html(website, header=0)[0]

In [94]:
print(CanPostal.shape)
CanPostal.columns

(180, 3)


Index(['Postal Code', 'Borough', 'Neighbourhood'], dtype='object')

In [95]:
CanPostal.dtypes



Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

### Examining rows with not assigned, and removing rows with Boroughs that aren't assigned

In [96]:
# Examining rows not assigned... 
CanPostalBNA = CanPostal[CanPostal['Borough'] == 'Not assigned']
#CanPostalBNA

As per instructions, removing rows that have borough that aren't assigned, and exmining where neighbourhood isn't assigned.

In [97]:
CanPostal.drop(CanPostal[CanPostal.Borough == 'Not assigned'].index, inplace=True)
CanPostal.shape

(103, 3)

In [98]:
CanPostalNNA = CanPostal[CanPostal['Neighbourhood'] == 'Not assigned']
#CanPostalNNA
# there are no neighbourhoods with "Not assigned" after removing others.

In [99]:
# Grouping by postal code first
CanPostal = CanPostal.groupby(['Postal Code', 'Borough'])['Neighbourhood'].apply(', '.join).reset_index()
print(CanPostal.shape)
CanPostal.head()

(103, 3)


Postal Code      Borough                           Neighbourhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

## Output of data scraped is 179 total postal codes
## However, 77 rows have Borough of 'Not assigned'.  These were removed. In addition, no other rows have neighborhoods that aren't assigned so no action is needed on neighborhoods.  The final number of 103 postal codes in Toronto match the website figure at the top of the table.

In [100]:
CanPostal.shape
print("Final Shape: ",CanPostal.shape)

Final Shape:  (103, 3)


# Week 3:  Section 2

In [101]:
# grabbing the Geocodes from file provided

coordfile = 'http://cocl.us/Geospatial_data'
cf=pd.read_csv(coordfile,header=0)
cf.sort_values(by='Postal Code', ascending=True)
cf.head(10)


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476
5         M1J  43.744734 -79.239476
6         M1K  43.727929 -79.262029
7         M1L  43.711112 -79.284577
8         M1M  43.716316 -79.239476
9         M1N  43.692657 -79.264848

In [102]:
cf.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [103]:
cf.shape

(103, 3)

In [104]:
j=CanPostal.merge(cf, on='Postal Code')
j

Postal Code           Borough  \
0           M1B       Scarborough   
1           M1C       Scarborough   
2           M1E       Scarborough   
3           M1G       Scarborough   
4           M1H       Scarborough   
5           M1J       Scarborough   
6           M1K       Scarborough   
7           M1L       Scarborough   
8           M1M       Scarborough   
9           M1N       Scarborough   
10          M1P       Scarborough   
11          M1R       Scarborough   
12          M1S       Scarborough   
13          M1T       Scarborough   
14          M1V       Scarborough   
15          M1W       Scarborough   
16          M1X       Scarborough   
17          M2H        North York   
18          M2J        North York   
19          M2K        North York   
20          M2L        North York   
21          M2M        North York   
22          M2N        North York   
23          M2P        North York   
24          M2R        North York   
25          M3A        North York   
26          M3B        North York   
27          M3C        North York   
28          M3H        North York   
29          M3J        North York   
30          M3K        North York   
31          M3L        North York   
32          M3M        North York   
33          M3N        North York   
34          M4A        North York   
35          M4B         East York   
36          M4C         East York   
37          M4E      East Toronto   
38          M4G         East York   
39          M4H         East York   
40          M4J         East York   
41          M4K      East Toronto   
42          M4L      East Toronto   
43          M4M      East Toronto   
44          M4N   Central Toronto   
45          M4P   Central Toronto   
46          M4R   Central Toronto   
47          M4S   Central Toronto   
48          M4T   Central Toronto   
49          M4V   Central Toronto   
50          M4W  Downtown Toronto   
51          M4X  Downtown Toronto   
52          M4Y  Downtown Toronto   
53          M5A  Downtown Toronto   
54          M5B  Downtown Toronto   
55          M5C  Downtown Toronto   
56          M5E  Downtown Toronto   
57          M5G  Downtown Toronto   
58          M5H  Downtown Toronto   
59          M5J  Downtown Toronto   
60          M5K  Downtown Toronto   
61          M5L  Downtown Toronto   
62          M5M        North York   
63          M5N   Central Toronto   
64          M5P   Central Toronto   
65          M5R   Central Toronto   
66          M5S  Downtown Toronto   
67          M5T  Downtown Toronto   
68          M5V  Downtown Toronto   
69          M5W  Downtown Toronto   
70          M5X  Downtown Toronto   
71          M6A        North York   
72          M6B        North York   
73          M6C              York   
74          M6E              York   
75          M6G  Downtown Toronto   
76          M6H      West Toronto   
77          M6J      West Toronto   
78          M6K      West Toronto   
79          M6L        North York   
80          M6M              York   
81          M6N              York   
82          M6P      West Toronto   
83          M6R      West Toronto   
84          M6S      West Toronto   
85          M7A  Downtown Toronto   
86          M7R       Mississauga   
87          M7Y      East Toronto   
88          M8V         Etobicoke   
89          M8W         Etobicoke   
90          M8X         Etobicoke   
91          M8Y         Etobicoke   
92          M8Z         Etobicoke   
93          M9A         Etobicoke   
94          M9B         Etobicoke   
95          M9C         Etobicoke   
96          M9L        North York   
97          M9M        North York   
98          M9N              York   
99          M9P         Etobicoke   
100         M9R         Etobicoke   
101         M9V         Etobicoke   
102         M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                       Malvern, Rouge  43.806686 -79.194353

In [105]:
#Final shape (section 2):
j.shape

(103, 5)

# Week 3:  Section 3

In [106]:
# Importing libraries:

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files


from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')








Libraries imported.


### Identifying and mapping the neighbourhoods in Toronto

In [107]:
address = 'Toronto, TO'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))


The geograpical coordinate of Toronto are 43.65238435, -79.38356765.


In [108]:

# create map of Toronto 
mapT = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighbourhood in zip(j['Latitude'], j['Longitude'], j['Borough'], j['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(mapT)  
    
mapT

#### Defining Foursquare Credentials and Version

In [109]:
CLIENT_ID = '1CETT04VB4CLSAGO1IXLCHKKZHVZRSETNG0TXNWYPDA4SEY0' # your Foursquare ID
CLIENT_SECRET = 'Y2GSJIWDGC45VYYYGMV0ASLAC5FI21JECGRODK5GIRC31C0K' # your Foursquare Secret
ACCESS_TOKEN = 'RR1BSSE2OGRMAEYF3YDUYQ0AIDNGDXNM10VZ04ZZZ3DGLQYZ' # your FourSquare Access Token
VERSION = '20200605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)
print('ACCESS_TOKEN:' + ACCESS_TOKEN)

Your credentails:
CLIENT_ID: 1CETT04VB4CLSAGO1IXLCHKKZHVZRSETNG0TXNWYPDA4SEY0
CLIENT_SECRET:Y2GSJIWDGC45VYYYGMV0ASLAC5FI21JECGRODK5GIRC31C0K
ACCESS_TOKEN:RR1BSSE2OGRMAEYF3YDUYQ0AIDNGDXNM10VZ04ZZZ3DGLQYZ


In [110]:
j.head()

Postal Code      Borough                           Neighbourhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

#### Looking at the first neighbourhood.



In [111]:
j.loc[0, 'Neighbourhood']

'Malvern, Rouge'

In [112]:
neighborhood_latitude = j.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = j.loc[0, 'Longitude'] # neighborhood longitude value
neighborhood_name = j.loc[0, 'Neighbourhood'] # neighborhood name
print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                              neighborhood_longitude))

Latitude and longitude values of Malvern, Rouge are 43.806686299999996, -79.19435340000001.


In [113]:
latitude = neighborhood_latitude 
longitude = neighborhood_longitude 
radius = 500
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&oauth_token={}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude,ACCESS_TOKEN, VERSION, radius, LIMIT)
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=1CETT04VB4CLSAGO1IXLCHKKZHVZRSETNG0TXNWYPDA4SEY0&client_secret=Y2GSJIWDGC45VYYYGMV0ASLAC5FI21JECGRODK5GIRC31C0K&v=20200605&ll=43.806686299999996,-79.19435340000001&radius=500&limit=100'

In [114]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '601c000aa2c7dd7e22585713'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4bb6b9446edc76b0d771311c-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/fastfood_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d16e941735',
         'name': 'Fast Food Restaurant',
         'pluralName': 'Fast Food Restaurants',
         'primary': True,
         'shortName': 'Fast Food'}],
       'id': '4bb6b9446edc76b0d771311c',
       'location': {'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'crossStreet': 'Morningside & Sheppard',
        'distance': 387,
        'formattedAddress': ['Toronto ON', 'Canada'],
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.80744841934756,
          'ln

#### Functions to loop through all venues in 1 neighborhood (Parkwood)

In [115]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [116]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name            categories        lat        lng
0  Wendy’s  Fast Food Restaurant  43.807448 -79.199056

In [117]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

1 venues were returned by Foursquare.


#### Functions to loop through all neighborhoods

In [118]:
# modified to include postal code and borough
def getNearbyVenues(names, postal, borough, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, post, bor, lat, lng in zip(names, postal, borough, latitudes, longitudes):
        print(name, post, bor)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()['response']['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            post,
            bor,
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Postal Code', 
                  'Borough', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [119]:
j.head()

Postal Code      Borough                           Neighbourhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [120]:
# Examining those with only Toronto in neighborhood name, as suggested.

In [121]:
CanPostalTorontoOnly= j[j['Borough'].str.contains("Toronto")]
print(CanPostalTorontoOnly.shape)
CanPostalTorontoOnly

(39, 5)


Postal Code           Borough  \
37         M4E      East Toronto   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A  Downtown Toronto   
87         M7Y      East Toronto   

                                        Neighbourhood   Latitude  Longitude  
37                                        The Beaches  43.676357 -79.293031  
41                       The Danforth West, Riverdale  43.679557 -79.352188  
42                     India Bazaar, The Beaches West  43.668999 -79.315572  
43                                    Studio District  43.659526 -79.340923  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
46                  North Toronto West, Lawrence Park  43.715383 -79.405678  
47                                         Davisville  43.704324 -79.388790  
48                        Moore Park, Summerhill East  43.689574 -79.383160  
49  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
50                                           Rosedale  43.679563 -79.377529  
51                        St. James Town, Cabbagetown  43.667967 -79.367675  
52                               Church and Wellesley  43.665860 -79.383160  
53                          Regent Park, Harbourfront  43.654260 -79.360636  
54                           Garden District, Ryerson  43.657162 -79.378937  
55                                     St. James Town  43.651494 -79.375418  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                           Richmond, Adelaide, King  43.650571 -79.384568  
59  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
60           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
61                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
63                                           Roselawn  43.711695 -79.416936  
64    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
65                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
66                     University of Toronto, Harbord  43.662696 -79.400049  
67          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
68  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
69                                     Stn A PO Boxes  43.646435 -79.374846  
70             First Canadian Place, Underground city  43.648429 -79.382280  
75                                           Christie  43.669542 -79.422564  
76                       Dufferin, Dovercourt Village  43.669005 -

## As suggested, I decided to look just at the Boroughs with  Toronto in the name (option was given in task). 
### I had run all, but was interested in looking at just the Toronto boroughs, so the final analysis is restricted to Toronto boroughs.

In [122]:
j=CanPostalTorontoOnly

In [123]:
print(j.shape)
j.head()

(39, 5)


Postal Code          Borough                   Neighbourhood   Latitude  \
37         M4E     East Toronto                     The Beaches  43.676357   
41         M4K     East Toronto    The Danforth West, Riverdale  43.679557   
42         M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
43         M4M     East Toronto                 Studio District  43.659526   
44         M4N  Central Toronto                   Lawrence Park  43.728020   

    Longitude  
37 -79.293031  
41 -79.352188  
42 -79.315572  
43 -79.340923  
44 -79.388790

In [124]:
# Added postal code and borough also to the dataframe
Tor_venues = getNearbyVenues(names=j['Neighbourhood'],postal=j['Postal Code'],borough=j['Borough'],latitudes=j['Latitude'],longitudes=j['Longitude'])

The Beaches M4E East Toronto
The Danforth West, Riverdale M4K East Toronto
India Bazaar, The Beaches West M4L East Toronto
Studio District M4M East Toronto
Lawrence Park M4N Central Toronto
Davisville North M4P Central Toronto
North Toronto West, Lawrence Park M4R Central Toronto
Davisville M4S Central Toronto
Moore Park, Summerhill East M4T Central Toronto
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park M4V Central Toronto
Rosedale M4W Downtown Toronto
St. James Town, Cabbagetown M4X Downtown Toronto
Church and Wellesley M4Y Downtown Toronto
Regent Park, Harbourfront M5A Downtown Toronto
Garden District, Ryerson M5B Downtown Toronto
St. James Town M5C Downtown Toronto
Berczy Park M5E Downtown Toronto
Central Bay Street M5G Downtown Toronto
Richmond, Adelaide, King M5H Downtown Toronto
Harbourfront East, Union Station, Toronto Islands M5J Downtown Toronto
Toronto Dominion Centre, Design Exchange M5K Downtown Toronto
Commerce Court, Victoria Hotel M5L Downtown Toronto


In [125]:
Tor_venues.head()

Neighborhood Postal Code       Borough  Neighborhood Latitude  \
0  The Beaches         M4E  East Toronto              43.676357   
1  The Beaches         M4E  East Toronto              43.676357   
2  The Beaches         M4E  East Toronto              43.676357   
3  The Beaches         M4E  East Toronto              43.676357   
4  The Beaches         M4E  East Toronto              43.676357   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.293031                 Seaspray Restaurant       43.678888   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.298167   Asian Restaurant

In [126]:
print(Tor_venues.shape)

(1589, 9)


In [127]:
print('There are {} uniques categories.'.format(len(Tor_venues['Venue Category'].unique())))

There are 231 uniques categories.


#### I had to replace "Neighborhood" with another title (now "Neighborhood Area"), as it's a designated neighborhood of interest in Foursquare

In [128]:
new_df = Tor_venues[Tor_venues['Venue Category'].str.contains('Neighborhood', na=False)]
new_df.head()

Neighborhood Postal Code  \
3                                          The Beaches         M4E   
75                                     Studio District         M4M   
664                           Richmond, Adelaide, King         M5H   
749  Harbourfront East, Union Station, Toronto Islands         M5J   

              Borough  Neighborhood Latitude  Neighborhood Longitude  \
3        East Toronto              43.676357              -79.293031   
75       East Toronto              43.659526              -79.340923   
664  Downtown Toronto              43.650571              -79.384568   
749  Downtown Toronto              43.640816              -79.381752   

                Venue  Venue Latitude  Venue Longitude Venue Category  
3       Upper Beaches       43.680563       -79.292869   Neighborhood  
75        Leslieville       43.662070       -79.337856   Neighborhood  
664  Downtown Toronto       43.653232       -79.385296   Neighborhood  
749      Harbourfront       43.639526       -79.380688   Neighborhood

In [129]:
Tor_venues['Venue Category']= Tor_venues['Venue Category'].replace("Neighborhood", "Neighborhood Area") 

## 3. Analyze Each Neighborhood

In [130]:
# one hot encoding
Tor_onehot = pd.get_dummies(Tor_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Tor_onehot['Neighborhood'] = Tor_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Tor_onehot.columns[-1]] + list(Tor_onehot.columns[:-1])
Tor_onehot = Tor_onehot[fixed_columns]

Tor_onehot.head()

Neighborhood  Adult Boutique  Airport  Airport Food Court  Airport Lounge  \
0  The Beaches               0        0                   0               0   
1  The Beaches               0        0                   0               0   
2  The Beaches               0        0                   0               0   
3  The Beaches               0        0                   0               0   
4  The Beaches               0        0                   0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 1   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       0     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Line  Business Service  Butcher  Café  \
0              0         0                 0        0     0   
1              0         0                 0        0     0   
2              0         0                 0        0     0   
3              0         0                 0        0     0   
4              0         0                 0        0     0   

   Cajun / Creole Restaurant  Candy Store  Ca

In [131]:
j_grouped=j.groupby(j['Neighbourhood']).mean().reset_index()
j_grouped

Neighbourhood   Latitude  Longitude
0                                         Berczy Park  43.644771 -79.373306
1        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191
2   Business reply mail Processing Centre, South C...  43.662744 -79.321558
3   CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420
4                                  Central Bay Street  43.657952 -79.387383
5                                            Christie  43.669542 -79.422564
6                                Church and Wellesley  43.665860 -79.383160
7                      Commerce Court, Victoria Hotel  43.648198 -79.379817
8                                          Davisville  43.704324 -79.388790
9                                    Davisville North  43.712751 -79.390197
10                       Dufferin, Dovercourt Village  43.669005 -79.442259
11             First Canadian Place, Underground city  43.648429 -79.382280
12    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307
13                           Garden District, Ryerson  43.657162 -79.378937
14  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752
15                      High Park, The Junction South  43.661608 -79.464763
16                     India Bazaar, The Beaches West  43.668999 -79.315572
17          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049
18                                      Lawrence Park  43.728020 -79.388790
19                           Little Portugal, Trinity  43.647927 -79.419750
20                        Moore Park, Summerhill East  43.689574 -79.383160
21                  North Toronto West, Lawrence Park  43.715383 -79.405678
22                             Parkdale, Roncesvalles  43.648960 -79.456325
23        Queen's Park, Ontario Provincial Government  43.662301 -79.389494
24                          Regent Park, Harbourfront  43.654260 -79.360636
25                           Richmond, Adelaide, King  43.650571 -79.384568
26                                           Rosedale  43.679563 -79.377529
27                                           Roselawn  43.711695 -79.416936
28                                 Runnymede, Swansea  43.651571 -79.484450
29                                     St. James Town  43.651494 -79.375418
30                        St. James Town, Cabbagetown  43.667967 -79.367675
31                                     Stn A PO Boxes  43.646435 -79.374846
32                                    Studio District  43.659526 -79.340923
33  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049
34                The Annex, North Midtown, Yorkville  43.672710 -79.405678
35                                        The Beaches  43.676357 -79.293031
36                       The Danforth West, Riverdale  43.679557 -79.352188
37           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576
38                     University of Toronto, Harbord  43.662696 -79.400049

In [132]:

Tor_grouped = Tor_onehot.groupby(Tor_onehot['Neighborhood']).mean().reset_index()
Tor_grouped

Neighborhood  Adult Boutique  \
0                                         Berczy Park        0.000000   
1        Brockton, Parkdale Village, Exhibition Place        0.000000   
2   Business reply mail Processing Centre, South C...        0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...        0.000000   
4                                  Central Bay Street        0.000000   
5                                            Christie        0.000000   
6                                Church and Wellesley        0.012987   
7                      Commerce Court, Victoria Hotel        0.000000   
8                                          Davisville        0.000000   
9                                    Davisville North        0.000000   
10                       Dufferin, Dovercourt Village        0.000000   
11             First Canadian Place, Underground city        0.000000   
12    Forest Hill North & West, Forest Hill Road Park        0.000000   
13                           Garden District, Ryerson        0.000000   
14  Harbourfront East, Union Station, Toronto Islands        0.000000   
15                      High Park, The Junction South        0.000000   
16                     India Bazaar, The Beaches West        0.000000   
17          Kensington Market, Chinatown, Grange Park        0.000000   
18                                      Lawrence Park        0.000000   
19                           Little Portugal, Trinity        0.000000   
20                        Moore Park, Summerhill East        0.000000   
21                  North Toronto West, Lawrence Park        0.000000   
22                             Parkdale, Roncesvalles        0.000000   
23        Queen's Park, Ontario Provincial Government        0.000000   
24                          Regent Park, Harbourfront        0.000000   
25                           Richmond, Adelaide, King        0.000000   
26                                           Rosedale        0.000000   
27                                           Roselawn        0.000000   
28                                 Runnymede, Swansea        0.000000   
29                                     St. James Town        0.000000   
30                        St. James Town, Cabbagetown        0.000000   
31                                     Stn A PO Boxes        0.000000   
32                                    Studio District        0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...        0.000000   
34                The Annex, North Midtown, Yorkville        0.000000   
35                                        The Beaches        0.000000   
36                       The Danforth West, Riverdale        0.000000   
37           Toronto Dominion Centre, Design Exchange        0.000000   
38                     University of Toronto, Harbord        0.000000   

     Airport  Airport Food Court  Airport Lounge  Airport Service  \
0   0.000000            0.000000        0.000000         0.000000   
1   0.000000            0.000000        0.000000         0.000000   
2   0.000000            0.000000        0.000000         0.000000   
3   0.076923            0.076923        0.076923         0.230769   
4   0.000000            0.000000        0.000000         0.000000   
5   0.000000            0.000000        0.000000         0.000000   
6   0.000000            0.000000        0.000000         0.000000   
7   0.000000            0.000000        0.000000         0.000000   
8   0.000000            0.000000        0.000000         0.000000   
9   0.000000            0.000000        0.000000         0.000000   
10  0.000000            0.000000        0.000000         0.000000   
11  0.000000            0.000000        0.000000         0.000000   
12  0.000000            0.000000        0.000000         0.000000   
13  0.000000            0.000000        0.000000         0.000000   
14  0.000000            0.000000        0.000000         0.000000   
15  0.000000    

In [133]:
Tor_grouped.shape

(39, 232)

In [134]:
num_top_venues = 10

for hood in Tor_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Tor_grouped[Tor_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.06
2            Beer Bar  0.04
3      Farmers Market  0.04
4         Cheese Shop  0.04
5          Restaurant  0.04
6              Bakery  0.04
7  Seafood Restaurant  0.04
8              Museum  0.02
9          Steakhouse  0.02


----Brockton, Parkdale Village, Exhibition Place----
                    venue  freq
0                    Café  0.13
1          Breakfast Spot  0.09
2             Coffee Shop  0.09
3           Grocery Store  0.04
4                  Bakery  0.04
5           Burrito Place  0.04
6            Climbing Gym  0.04
7       Convenience Store  0.04
8  Furniture / Home Store  0.04
9                     Gym  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                  venue  freq
0    Light Rail Station  0.12
1           Pizza Place  0.06
2         Auto Workshop  0.06
3                  Park  0.06
4            Restaur

### Convert into a pandas Dataframe

In [135]:
# Function to sort venues in decending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [136]:
Tor_grouped.shape

(39, 232)

In [137]:
# get dbs of top 10 venues per neighborhood

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Tor_grouped['Neighborhood']

for ind in np.arange(Tor_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Tor_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar    Seafood Restaurant            Restaurant   
1        Breakfast Spot           Coffee Shop           Yoga Studio   
2           Pizza Place        Farmers Market            Skate Park   
3      Airport Terminal         Boat or Ferry   Rental Car Location   
4    Italian Restaurant        Sandwich Place                  Café   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0                Bakery              Beer Bar         Farmers Market   
1                   Gym             Pet Store  Performing Arts Venue   
2                  Park            Comic Shop   Gym / Fitness Center   
3      Sculpture Garden              Boutique        Harbor / Marina   
4           Salad Place       Bubble Tea Shop           Burger Joint   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Cheese Shop                 Beach                 Bistro  
1                  Nightclub    Italian Restaurant           Intersection  
2                 Smoke Shop         Garden Center                 Garden  
3             Airport Lounge    Airport Food Court                Airport  
4  Middle Eastern Restaurant      Ramen Restaurant  Portuguese Restaurant

In [138]:
neighborhoods_venues_sorted.shape

(39, 11)

### Clustering Neighborhoods using k-means 

In [139]:
# set number of clusters
kclusters = 5

Tor_grouped_clustering = Tor_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Tor_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [140]:
neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar    Seafood Restaurant            Restaurant   
1        Breakfast Spot           Coffee Shop           Yoga Studio   
2           Pizza Place        Farmers Market            Skate Park   
3      Airport Terminal         Boat or Ferry   Rental Car Location   
4    Italian Restaurant        Sandwich Place                  Café   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0                Bakery              Beer Bar         Farmers Market   
1                   Gym             Pet Store  Performing Arts Venue   
2                  Park            Comic Shop   Gym / Fitness Center   
3      Sculpture Garden              Boutique        Harbor / Marina   
4           Salad Place       Bubble Tea Shop           Burger Joint   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Cheese Shop                 Beach                 Bistro  
1                  Nightclub    Italian Restaurant           Intersection  
2                 Smoke Shop         Garden Center                 Garden  
3             Airport Lounge    Airport Food Court                Airport  
4  Middle Eastern Restaurant      Ramen Restaurant  Portuguese Restaurant

In [141]:
j_grouped = j_grouped.rename(columns={'Neighbourhood': 'Neighborhood'})
print(j_grouped.shape)
j_grouped.head()

(39, 3)


Neighborhood   Latitude  Longitude
0                                        Berczy Park  43.644771 -79.373306
1       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191
2  Business reply mail Processing Centre, South C...  43.662744 -79.321558
3  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420
4                                 Central Bay Street  43.657952 -79.387383

In [142]:
print(neighborhoods_venues_sorted.shape)
neighborhoods_venues_sorted.head()

(39, 11)


Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar    Seafood Restaurant            Restaurant   
1        Breakfast Spot           Coffee Shop           Yoga Studio   
2           Pizza Place        Farmers Market            Skate Park   
3      Airport Terminal         Boat or Ferry   Rental Car Location   
4    Italian Restaurant        Sandwich Place                  Café   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0                Bakery              Beer Bar         Farmers Market   
1                   Gym             Pet Store  Performing Arts Venue   
2                  Park            Comic Shop   Gym / Fitness Center   
3      Sculpture Garden              Boutique        Harbor / Marina   
4           Salad Place       Bubble Tea Shop           Burger Joint   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Cheese Shop                 Beach                 Bistro  
1                  Nightclub    Italian Restaurant           Intersection  
2                 Smoke Shop         Garden Center                 Garden  
3             Airport Lounge    Airport Food Court                Airport  
4  Middle Eastern Restaurant      Ramen Restaurant  Portuguese Restaurant

In [143]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [144]:
neighborhoods_venues_sorted.head()

Cluster Labels                                       Neighborhood  \
0               0                                        Berczy Park   
1               0       Brockton, Parkdale Village, Exhibition Place   
2               0  Business reply mail Processing Centre, South C...   
3               0  CN Tower, King and Spadina, Railway Lands, Har...   
4               0                                 Central Bay Street   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop          Cocktail Bar    Seafood Restaurant   
1                  Café        Breakfast Spot           Coffee Shop   
2    Light Rail Station           Pizza Place        Farmers Market   
3       Airport Service      Airport Terminal         Boat or Ferry   
4           Coffee Shop    Italian Restaurant        Sandwich Place   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0            Restaurant                Bakery              Beer Bar   
1           Yoga Studio                   Gym             Pet Store   
2            Skate Park                  Park            Comic Shop   
3   Rental Car Location      Sculpture Garden              Boutique   
4                  Café           Salad Place       Bubble Tea Shop   

   7th Most Common Venue      8th Most Common Venue 9th Most Common Venue  \
0         Farmers Market                Cheese Shop                 Beach   
1  Performing Arts Venue                  Nightclub    Italian Restaurant   
2   Gym / Fitness Center                 Smoke Shop         Garden Center   
3        Harbor / Marina             Airport Lounge    Airport Food Court   
4           Burger Joint  Middle Eastern Restaurant      Ramen Restaurant   

  10th Most Common Venue  
0                 Bistro  
1           Intersection  
2                 Garden  
3                Airport  
4  Portuguese Restaurant

In [145]:
# join
Tor_merged = j_grouped

# merge Tor_grouped with j to add latitude/longitude for each neighborhood

#Tor_merged = Tor_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood',how='left')
Tor_merged = Tor_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Tor_merged.head() # check the last columns!

Neighborhood   Latitude  Longitude  \
0                                        Berczy Park  43.644771 -79.373306   
1       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
2  Business reply mail Processing Centre, South C...  43.662744 -79.321558   
3  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420   
4                                 Central Bay Street  43.657952 -79.387383   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0           Coffee Shop          Cocktail Bar   
1               0                  Café        Breakfast Spot   
2               0    Light Rail Station           Pizza Place   
3               0       Airport Service      Airport Terminal   
4               0           Coffee Shop    Italian Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0    Seafood Restaurant            Restaurant                Bakery   
1           Coffee Shop           Yoga Studio                   Gym   
2        Farmers Market            Skate Park                  Park   
3         Boat or Ferry   Rental Car Location      Sculpture Garden   
4        Sandwich Place                  Café           Salad Place   

  6th Most Common Venue  7th Most Common Venue      8th Most Common Venue  \
0              Beer Bar         Farmers Market                Cheese Shop   
1             Pet Store  Performing Arts Venue                  Nightclub   
2            Comic Shop   Gym / Fitness Center                 Smoke Shop   
3              Boutique        Harbor / Marina             Airport Lounge   
4       Bubble Tea Shop           Burger Joint  Middle Eastern Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0                 Beach                 Bistro  
1    Italian Restaurant           Intersection  
2         Garden Center                 Garden  
3    Airport Food Court                Airport  
4      Ramen Restaurant  Portuguese Restaurant

In [146]:
Tor_merged.columns

Index(['Neighborhood', 'Latitude', 'Longitude', 'Cluster Labels',
       '1st Most Common Venue', '2nd Most Common Venue',
       '3rd Most Common Venue', '4th Most Common Venue',
       '5th Most Common Venue', '6th Most Common Venue',
       '7th Most Common Venue', '8th Most Common Venue',
       '9th Most Common Venue', '10th Most Common Venue'],
      dtype='object')

In [147]:
Tor_merged=Tor_merged.dropna()
Tor_merged.reset_index()
Tor_merged.head()

Neighborhood   Latitude  Longitude  \
0                                        Berczy Park  43.644771 -79.373306   
1       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
2  Business reply mail Processing Centre, South C...  43.662744 -79.321558   
3  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420   
4                                 Central Bay Street  43.657952 -79.387383   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0           Coffee Shop          Cocktail Bar   
1               0                  Café        Breakfast Spot   
2               0    Light Rail Station           Pizza Place   
3               0       Airport Service      Airport Terminal   
4               0           Coffee Shop    Italian Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0    Seafood Restaurant            Restaurant                Bakery   
1           Coffee Shop           Yoga Studio                   Gym   
2        Farmers Market            Skate Park                  Park   
3         Boat or Ferry   Rental Car Location      Sculpture Garden   
4        Sandwich Place                  Café           Salad Place   

  6th Most Common Venue  7th Most Common Venue      8th Most Common Venue  \
0              Beer Bar         Farmers Market                Cheese Shop   
1             Pet Store  Performing Arts Venue                  Nightclub   
2            Comic Shop   Gym / Fitness Center                 Smoke Shop   
3              Boutique        Harbor / Marina             Airport Lounge   
4       Bubble Tea Shop           Burger Joint  Middle Eastern Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0                 Beach                 Bistro  
1    Italian Restaurant           Intersection  
2         Garden Center                 Garden  
3    Airport Food Court                Airport  
4      Ramen Restaurant  Portuguese Restaurant

In [148]:
Tor_merged.head()

Neighborhood   Latitude  Longitude  \
0                                        Berczy Park  43.644771 -79.373306   
1       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
2  Business reply mail Processing Centre, South C...  43.662744 -79.321558   
3  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420   
4                                 Central Bay Street  43.657952 -79.387383   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0           Coffee Shop          Cocktail Bar   
1               0                  Café        Breakfast Spot   
2               0    Light Rail Station           Pizza Place   
3               0       Airport Service      Airport Terminal   
4               0           Coffee Shop    Italian Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0    Seafood Restaurant            Restaurant                Bakery   
1           Coffee Shop           Yoga Studio                   Gym   
2        Farmers Market            Skate Park                  Park   
3         Boat or Ferry   Rental Car Location      Sculpture Garden   
4        Sandwich Place                  Café           Salad Place   

  6th Most Common Venue  7th Most Common Venue      8th Most Common Venue  \
0              Beer Bar         Farmers Market                Cheese Shop   
1             Pet Store  Performing Arts Venue                  Nightclub   
2            Comic Shop   Gym / Fitness Center                 Smoke Shop   
3              Boutique        Harbor / Marina             Airport Lounge   
4       Bubble Tea Shop           Burger Joint  Middle Eastern Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0                 Beach                 Bistro  
1    Italian Restaurant           Intersection  
2         Garden Center                 Garden  
3    Airport Food Court                Airport  
4      Ramen Restaurant  Portuguese Restaurant

In [149]:
Tor_merged.dtypes
Tor_merged['Cluster Labels']=Tor_merged['Cluster Labels'].astype('int64') 
#df['cluster_labels']=df['cluster_labels'].astype('int64') 
Tor_merged

Neighborhood   Latitude  Longitude  \
0                                         Berczy Park  43.644771 -79.373306   
1        Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
2   Business reply mail Processing Centre, South C...  43.662744 -79.321558   
3   CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420   
4                                  Central Bay Street  43.657952 -79.387383   
5                                            Christie  43.669542 -79.422564   
6                                Church and Wellesley  43.665860 -79.383160   
7                      Commerce Court, Victoria Hotel  43.648198 -79.379817   
8                                          Davisville  43.704324 -79.388790   
9                                    Davisville North  43.712751 -79.390197   
10                       Dufferin, Dovercourt Village  43.669005 -79.442259   
11             First Canadian Place, Underground city  43.648429 -79.382280   
12    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307   
13                           Garden District, Ryerson  43.657162 -79.378937   
14  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
15                      High Park, The Junction South  43.661608 -79.464763   
16                     India Bazaar, The Beaches West  43.668999 -79.315572   
17          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
18                                      Lawrence Park  43.728020 -79.388790   
19                           Little Portugal, Trinity  43.647927 -79.419750   
20                        Moore Park, Summerhill East  43.689574 -79.383160   
21                  North Toronto West, Lawrence Park  43.715383 -79.405678   
22                             Parkdale, Roncesvalles  43.648960 -79.456325   
23        Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
24                          Regent Park, Harbourfront  43.654260 -79.360636   
25                           Richmond, Adelaide, King  43.650571 -79.384568   
26                                           Rosedale  43.679563 -79.377529   
27                                           Roselawn  43.711695 -79.416936   
28                                 Runnymede, Swansea  43.651571 -79.484450   
29                                     St. James Town  43.651494 -79.375418   
30                        St. James Town, Cabbagetown  43.667967 -79.367675   
31                                     Stn A PO Boxes  43.646435 -79.374846   
32                                    Studio District  43.659526 -79.340923   
33  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049   
34                The Annex, North Midtown, Yorkville  43.672710 -79.405678   
35                                        The Beaches  43.676357 -79.293031   
36                       The Danforth West, Riverdale  43.679557 -79.352188   
37           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
38                     University of Toronto, Harbord  43.662696 -79.400049   

    Cluster Labels 1st Most Common Venue          2nd Most Common Venue  \
0                0           Coffee Shop                   Cocktail Bar   
1                0                  Café                 Breakfast Spot   
2                0    Light Rail Station                    Pizza Place   
3                0       Airport Service               Airport Terminal   
4                0           Coffee Shop             Italian Restaurant   
5                0         Grocery Store                           Café   
6                0           Coffee Shop            Japanese Restaurant   
7                0           Coffee Shop                     Restaurant   
8                0        Sandwich Place                   Dessert Shop   
9                0                  Park                    Pizza Place   
10               0                Bakery               

In [150]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Tor_merged['Latitude'], Tor_merged['Longitude'], Tor_merged['Neighborhood'], Tor_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Analysis: Final clusters for all of the neighborhoods:  most clustered into 1 of the 5 clusters -- Cluster 1 was more cafes and restaurants recreational/sports while other clusters were more about recreation.  If I ran this again, I'd probably divide into 2 or 3 clusters, but they still would show cafes and restaurants as dominant.